presto-main/src/main/java/com/facebook/presto/operator/exchange/PartitioningExchanger.java

```diff
@@ -62,40 +63,61 @@ public PartitioningExchanger(
         }
         partitionGenerator = new LocalPartitionGenerator(hashGenerator, buffers.size());
 
-        partitionAssignments = new IntList[partitions.size()];
-        for (int i = 0; i < partitionAssignments.length; i++) {
-            partitionAssignments[i] = new IntArrayList();
-        }
+        pageBuilders = IntStream.range(0, buffers.size())
+                .mapToObj(i -> new PageBuilder(types))
+                .collect(toImmutableList());
     }
 
     @Override
-    public synchronized void accept(Page page)
+    public synchronized void exchange(Page page)
     {
-        // reset the assignment lists
-        for (IntList partitionAssignment : partitionAssignments) {
-            partitionAssignment.clear();
-        }
-
-        // assign each row to a partition
+        long initialRetainedSize = getRetainedSizeInBytes();
         for (int position = 0; position < page.getPositionCount(); position++) {
             int partition = partitionGenerator.getPartition(position, page);
-            partitionAssignments[partition].add(position);
+            PageBuilder pageBuilder = pageBuilders.get(partition);
+            appendRow(pageBuilder, page, position);
         }
+        flush(false);
+        long finalRetainedSize = getRetainedSizeInBytes();
+        memoryTracker.accept(finalRetainedSize - initialRetainedSize);
+    }
+
+    @Override
+    public void flush()
+    {
+        long initialRetainedSize = getRetainedSizeInBytes();
+        flush(true);
+        long finalRetainedSize = getRetainedSizeInBytes();
+        memoryTracker.accept(finalRetainedSize - initialRetainedSize);
+    }
 
-        // build a page for each partition
-        Block[] sourceBlocks = page.getBlocks();
-        Block[] outputBlocks = new Block[sourceBlocks.length];
-        for (int partition = 0; partition < buffers.size(); partition++) {
-            List<Integer> positions = partitionAssignments[partition];
-            if (!positions.isEmpty()) {
-                for (int i = 0; i < sourceBlocks.length; i++) {
-                    outputBlocks[i] = sourceBlocks[i].copyPositions(positions);
-                }
+    private void appendRow(PageBuilder pageBuilder, Page page, int position)
+    {
+        pageBuilder.declarePosition();
```

Comment:
> com.facebook.presto.spi.block.Block#writeBytesTo

I think i don't understand this method. `writeBytesTo` allows you to copy chunk of value in a position. Not the entire block. Am i wrong?